app/src/main/java/de/xikolo/storages/RecentCoursesStorage.kt

```diff
@@ -0,0 +1,55 @@
+package de.xikolo.storages
+
+import android.content.Context
+import androidx.preference.PreferenceManager
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import de.xikolo.App
+import de.xikolo.storages.base.BaseStorage
+import de.xikolo.storages.base.SharedPreferenceLiveData
+
+class RecentCoursesStorage : BaseStorage(PREF_RECENT_COURSES, Context.MODE_PRIVATE) {
+
+    val recentCourses: LinkedHashSet<Pair<String, String>>?
+        get() {
+            val json = getString(PREF_RECENT_COURSES)
+            val type = object : TypeToken<LinkedHashSet<Pair<String, String>>>() {}.type
+            return Gson().fromJson(json, type)
+        }
+
+    private val shortcutMax = 3
+    val coursesLive: SharedPreferenceLiveData<String> by lazy {
+        // what name does the preference have? What is the sharedPref Parameter supposed to do?
+        SharedPreferenceLiveData.SharedPreferenceStringLiveData(PreferenceManager.getDefaultSharedPreferences(App.instance), "recentCourses", "")
+    }
+
+    fun addCourse(courseId: String, title: String) {
+        var courses: LinkedHashSet<Pair<String, String>>? = recentCourses
+
+        if (recentCourses == null) {
+            courses = LinkedHashSet<Pair<String, String>>()
+        }
+
+        when {
+            courses!!.contains(Pair(courseId, title)) -> {
+                courses.remove(Pair(courseId, title))
+                courses.add(Pair(courseId, title))
+            }
+            courses.size < shortcutMax                -> {
+                courses.add(Pair(courseId, title))
+            }
+            courses.size == shortcutMax               -> {
+                courses.remove(courses.last())
+                courses.add(Pair(courseId, title))
+            }
+        }
+
+        putString(PREF_RECENT_COURSES, Gson().toJson(courses))
+        //is this update procedure correct?
+        coursesLive.update(Gson().toJson(courses))
```

Comment:
For the liveData class I created was no public setValue() available. So I wrote an update method, that calls the setValue(). Is this correct?

Reply:
You should update the preferences not the live data object.

Reply:
The live data object only wraps the preference and offers an observer if the pref value was changed.

Reply:
So these lines can be deleted. The preferences are updated in the line above.